File: samples/vision/lanenet/runtime/python/visualization.py
# Copyright (c) 2026 D-Robotics Corporation
# SPDX-License-Identifier: Apache-2.0
"""Source-native-style displays; colors do not represent clustered lane IDs."""

import numpy as np


def embedding_image(embedding):
    """Finite float32 CHW3 → HWC uint8 clip/round; channel order is preserved.

    This unifies Python with native clipping/saturating rounding and deliberately
    replaces source Python truncation/wrap. It is a display, not instance labels.
    """
    value = np.asarray(embedding)
    if (
        value.ndim != 3
        or value.shape[0] != 3
        or value.size == 0
        or value.dtype != np.float32
        or not np.isfinite(value).all()
    ):
        raise ValueError("Expected finite float32 CHW embedding with three channels")
    return np.rint(np.clip(value, 0, 1) * 255).transpose(1, 2, 0).astype(np.uint8)


def binary_image(binary):
    """2D discrete labels → 0/255 display; reject invalid labels."""
    value = np.asarray(binary)
    if (
        value.ndim != 2
        or value.size == 0
        or value.dtype.kind not in "iu"
        or not np.isin(value, (0, 1)).all()
    ):
        raise ValueError("Expected 2D integer labels 0/1")
    return value.astype(np.uint8) * 255

File: samples/vision/lanenet/runtime/cpp/src/visualization.cpp
// Copyright (c) 2026 D-Robotics Corporation
// SPDX-License-Identifier: Apache-2.0
#include "visualization.hpp"
#include <stdexcept>
namespace lanenet {
cv::Mat embedding_image(const LaneResult &result) {
  if (result.embedding.size() != 3 * 256 * 512)
    throw std::invalid_argument("Wrong embedding geometry");
  cv::Mat image(256, 512, CV_8UC3);
  for (int h = 0; h < 256; h++)
    for (int w = 0; w < 512; w++)
      for (int c = 0; c < 3; c++)
        image.at<cv::Vec3b>(h, w)[c] =
            display_component(result.embedding[c * 256 * 512 + h * 512 + w]);
  return image; // preserved channel order; not lane-instance IDs
}
cv::Mat binary_image(const LaneResult &result) {
  if (result.binary.size() != 256 * 512)
    throw std::invalid_argument("Wrong binary geometry");
  cv::Mat image(256, 512, CV_8UC1);
  for (int h = 0; h < 256; h++)
    for (int w = 0; w < 512; w++) {
      const auto value = result.binary[h * 512 + w];
      if (value > 1)
        throw std::invalid_argument("Invalid binary label");
      image.at<unsigned char>(h, w) = value * 255;
    }
  return image;
}
} // namespace lanenet
